src/lib.rs

```diff
@@ -411,6 +411,80 @@ impl<'text> BidiInfo<'text> {
         result.into()
     }
 
+    /// Returns a map from the visual order to the logical order.
+    ///
+    ///   ## Example
+    ///
+    /// ```rust
+    /// # #[cfg(feature = "hardcoded-data")] {
+    /// use unicode_bidi::BidiInfo;
+    ///
+    /// // This example text is defined using `concat!` because some browsers
+    /// // and text editors have trouble displaying bidi strings.
+    /// let text = concat![
+    ///   "א",
+    ///   "ב",
+    ///   "ג",
+    ///   "a",
+    ///   "b",
+    ///   "c",
+    /// ];
+    ///
+    /// // Resolve embedding levels within the text.  Pass `None` to detect the
+    /// // paragraph level automatically.
+    /// let bidi_info = BidiInfo::new(&text, None);
+    ///
+    /// // This paragraph has embedding level 1 because its first strong character is RTL.
+    /// assert_eq!(bidi_info.paragraphs.len(), 1);
+    /// let para = &bidi_info.paragraphs[0];
+    /// assert_eq!(para.level.number(), 1);
+    /// assert_eq!(para.level.is_rtl(), true);
+    ///
+    /// // Re-ordering is done after wrapping each paragraph into a sequence of
+    /// // lines. For this example, I'll just use a single line that spans the
+    /// // entire paragraph.
+    /// let line = para.range.clone();
+    ///
+    /// let display = bidi_info.reorder_line(para, line);
+    /// assert_eq!(display, concat![
+    ///   "a",
+    ///   "b",
+    ///   "c",
+    ///   "ג",
+    ///   "ב",
+    ///   "א",
+    /// ]);
+    ///
+    /// let line = para.range.clone();
+    /// let order = bidi_info.reorder_index(para);
+    ///
+    /// assert_eq!(order, [ 6, 7, 8, 4, 2, 0 ]);
+    /// assert_eq!(bidi_info.text[6..].chars().next().unwrap(), 'a');
+    /// assert_eq!(bidi_info.text[7..].chars().next().unwrap(), 'b');
+    /// assert_eq!(bidi_info.text[8..].chars().next().unwrap(), 'c');
+    /// assert_eq!(bidi_info.text[4..].chars().next().unwrap(), 'ג');
+    /// assert_eq!(bidi_info.text[2..].chars().next().unwrap(), 'ב');
+    /// assert_eq!(bidi_info.text[0..].chars().next().unwrap(), 'א');
+    /// # } // feature = "hardcoded-data"
+    ///
+    pub fn reorder_index(&self, para: &ParagraphInfo) -> Vec<usize> {
```

Comment:
@sffc does this sound like a good Rusty way to have this API?

I think it's mostly fine, we might need to wrap it over FFI

Reply:
I think it looks okay; it's not great that it returns a Vec, but most functions in this file already return Vec. I am not currently comfortable commenting on the utility of this function in terms of the needs of Skia.

Reply:
Should we wait to merge it until you have a better answer there?

Reply:
We may need a more step by step thing with the level map

Reply:
Of the three of us in this PR, I am the least familiar with the workings of the bi-di algorithm. I don't want to be the one making a recommendation on whether this is a positive change in the API surface of unicode_bidi.

Reply:
Oh, no, this question is based on the utility in terms of what Skia needs. I think this is a fine thing to add to unicode_bidi, I don't know if it's _sufficient_.

Reply:
I see. Yeah, I think it's correct, but it's not the exact same as `ubidi_reorderVisual`; the function proposed here is the analog of `ubidi_getVisualMap`.

Reply:
@sffc , Okay, in this PR, https://github.com/servo/unicode-bidi/pull/80 , I have created an exact match to `ubidi_reorderVisual`, However, I do not think it should be here, because it does not make sense to have this helper without a context. I suggest to move it as a helper in ICU4X.

For the current PR, I will mimic `ubidi_getVisualMap` and  `ubidi_getLogicalMap` because they are important for the users any way.